### ISSUE TITLE
Allow RegisterCollection to ignore composite types

### ISSUE THREAD

**Issue by dhilgarth** [simpleinjector/SimpleInjector#119]
I try to avoid working with enumerables of my interfaces where I just have to execute a method on each instance. Instead, I am using a composite, so that the caller just has to deal with one instance.
So, what I often do is this:

``` c#
container.RegisterCollection<IStartUpHandler>(relevantAssemblies);
container.RegisterSingleton<IStartUpHandler, CompositeStartUpHandler>();
```

The problem here is that `RegisterCollection` also registers the `CompositeStartUpHandler` which will result in a `CyclicDependency...Exception`, when trying to get an instance of `IStartUpHandler`.

Right now, I am using my own helper method that ignores composites:

``` c#
var behavior = container.Options.ConstructorResolutionBehavior;

var types = 
    from assembly in assemblies
    from type in assembly.GetTypes()
    where typeof(TService).IsAssignableFrom(type)
    where !type.IsAbstract
    where !type.IsGenericTypeDefinition
    where !IsCompositeOf<TService>(type, behavior)
    select type;

container.RegisterCollection<TService>(types);
```

With `IsCompositeOf` being implemented like this:

``` c#
static bool IsCompositeOf<TService>(Type type, IConstructorResolutionBehavior behavior)
{
    var ctor = behavior.GetConstructor(typeof(TService), type);

    return ctor.GetParameters().Any(p => p.ParameterType == typeof(IEnumerable<TService>));
}
```

It would be nice, if this would somehow make it into the core code-base.

**Comment by dotnetjunkie** [comment]
I've marked this as a bug. We overlooked this during the development of v3. We will fix this in the next patch release. Thanks for reporting this.

**Comment by dhilgarth** [comment]
You are welcome. Your response times are absolutely crazy! :-)

**Comment by dotnetjunkie** [comment]
We're in the same time zone buddy :) The real question here is, why are you posting bug reports on Github on 7:30 in the morning? :D

**Comment by dhilgarth** [comment]
What I meant was: You are reacting immediately. I have always been wondering, how you are able to perform these context switches so quickly from you work to the message back to your work.

As to why I post at 7:30: Well, I just encountered that issue, so why not post it while it's fresh? :)

**Comment by dotnetjunkie** [comment]
Don't worry, I was just joking. I'm not always able to make the context switch quickly, since I have to work, sleep, eat, sport, and have social life (or well, sort of). But when I can, I try to respond a.s.a.p. to all incoming questions and issues concerning Simple Injector.

**Comment by dotnetjunkie** [comment]
Do note that your helper code can be simplified to the following

``` c#
var types = 
    from type in container.GetTypesToRegister(typeof(TService), assemblies)
    where !IsCompositeOf<TService>(type, container.Options.ConstructorResolutionBehavior)
    select type;

container.RegisterCollection<TService>(types);
```

The `Container.GetTypesToRegister` method has been added in v3. This method is used internally by `Register` and `RegisterCollection` and retrieves all implementations of a given type (both open, closed and non-generic) and it already filters out decorators, abstract types and generic type definitions.

**Comment by dhilgarth** [comment]
Thanks for the hint!

**Comment by dotnetjunkie** [comment]
I moved this to milestone v3.1, because it is a little bit feature-ish and it becomes much easier to implement after #122 is implemented.

**Comment by dhilgarth** [comment]
Fine with me, I have the workaround in place anyway :)

**Comment by dhilgarth** [comment]
Thanks! :)

**Comment by dotnetjunkie** [comment]
We just released v3.1. In v3.1, `RegisterCollection` will automatically filter out composites.

**Comment by dhilgarth** [comment]
Awesome, thanks!
